kratos/tests/test_tensor_adaptors.py

```diff
@@ -0,0 +1,85 @@
+                Kratos.EXTERNAL_FORCES_VECTOR,  # vector
+                Kratos.NORMAL_SHAPE_DERIVATIVE  # matrix
+            ]
+
+        cls.output_list_of_variables = [
+                Kratos.DENSITY,                 # double
+                Kratos.ACCELERATION,            # array3
+                Kratos.INTERNAL_FORCES_VECTOR,  # vector
+                Kratos.PK2_STRESS_TENSOR        # matrix
+            ]
+
+        for var in cls.input_list_of_variables:
+            cls.model_part.AddNodalSolutionStepVariable(var)
+
+        for var in cls.output_list_of_variables:
+            cls.model_part.AddNodalSolutionStepVariable(var)
+
+        for i in range(10):
+            cls.model_part.CreateNewNode(i + 1, i, i + 1, i + 2)
+
+    def setUp(self):
+        for node in self.model_part.Nodes:
+            node.SetSolutionStepValue(Kratos.PRESSURE, node.Id + 1)
+            node.SetSolutionStepValue(Kratos.VELOCITY, [node.Id * 2 + 1, node.Id * 3 + 1, node.Id * 4 + 1])
+            node.SetSolutionStepValue(Kratos.EXTERNAL_FORCES_VECTOR, Kratos.Vector([node.Id * 2 + 1, node.Id * 3 + 1, node.Id * 4 + 1, node.Id * 2 + 1, node.Id * 3 + 1]))
+            node.SetSolutionStepValue(Kratos.NORMAL_SHAPE_DERIVATIVE, Kratos.Matrix([[node.Id * 2 + 1, node.Id * 3 + 1], [node.Id * 4 + 1, node.Id * 2 + 1], [node.Id * 3 + 1, node.Id * 4 + 1]]))
+
+    def test_NodeHistoricalVariableTensorAdaptorShape(self):
+        with self.assertRaises(RuntimeError):
+            Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, Kratos.VELOCITY, [10, 4], 0)
+            Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, Kratos.VELOCITY, [3, 3], 0)
+            Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, Kratos.VELOCITY, [3, 3, 3], 0)
+
+        self.assertEqual(Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, Kratos.EXTERNAL_FORCES_VECTOR, 0).Shape(), [10, 5])
+        self.assertEqual(Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, Kratos.INTERNAL_FORCES_VECTOR, 0).Shape(), [10, 0])
+        self.assertEqual(Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, Kratos.INTERNAL_FORCES_VECTOR, [10, 6], 0).Shape(), [10, 6])
+
+    def test_NodeHistoricalVariableTensorAdaptor(self):
+        for input_var, output_var in zip(self.input_list_of_variables, self.output_list_of_variables):
+            read_tensor_adaptor = Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(self.model_part.Nodes, input_var)
+            write_tensor_adaptor = Kratos.TensorAdaptors.NodeTensorAdaptors.NodeHistoricalVariableTensorAdaptor(read_tensor_adaptor.GetContainer(), output_var, read_tensor_adaptor.Shape(), 1)
```

Comment:
i would say that here the shape can be automatically computed. If you have the container and the variable  (output_var) than the shape of the container is known